<a href="https://colab.research.google.com/github/IngyinnK/My-Python-Pandas-Projects/blob/main/02_02_Pandas_Day_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
# install pandas library
!pip install pandas

In [ ]:
# Requirement already satisfied means library already installed.

In [ ]:
# import pandas library
import pandas as pd

In [ ]:
# check version
pd.__version__

'2.2.2'

# Import Dataset


In [ ]:
df = pd.read_csv("student.csv")
# If excel file, pd.read_excel("student.xlsx").
# df represents dataframe.
# create df variable and assign it to dataset.

In [ ]:
df

,id,name,class,mark,gender
0,1,John Deo,Four,75,female
1,2,Max Ruin,Three,85,male
2,3,Arnold,Three,55,male
3,4,Krish Star,Four,60,female
4,5,John Mike,Four,60,female
5,6,Alex John,Four,55,male
6,7,My John Rob,Fifth,78,male
7,8,Asruid,Five,85,male
8,9,Tes Qry,Six,78,NaN
9,10,Big John,Four,55,female


In [ ]:
# first 5 rows
df.head()

,id,name,class,mark,gender
0,1,John Deo,Four,75,female
1,2,Max Ruin,Three,85,male
2,3,Arnold,Three,55,male
3,4,Krish Star,Four,60,female
4,5,John Mike,Four,60,female


In [ ]:
# first 10 rows
df.head(10)

,id,name,class,mark,gender
0,1,John Deo,Four,75,female
1,2,Max Ruin,Three,85,male
2,3,Arnold,Three,55,male
3,4,Krish Star,Four,60,female
4,5,John Mike,Four,60,female
5,6,Alex John,Four,55,male
6,7,My John Rob,Fifth,78,male
7,8,Asruid,Five,85,male
8,9,Tes Qry,Six,78,NaN
9,10,Big John,Four,55,female


In [ ]:
# last 5 rows
df.tail()

,id,name,class,mark,gender
30,31,Marry Toeey,Four,88,male
31,32,Binn Rott,Seven,90,female
32,33,Kenn Rein,Six,96,female
33,34,Gain Toe,Seven,69,male
34,35,Rows Noump,Six,88,female


In [ ]:
# last 10 rows
df.tail(10)

,id,name,class,mark,gender
25,26,Crelea,Seven,79,male
26,27,NaN,Three,81,NaN
27,28,Rojj Base,Seven,86,female
28,29,Tess Played,Seven,55,male
29,30,Reppy Red,Six,79,female
30,31,Marry Toeey,Four,88,male
31,32,Binn Rott,Seven,90,female
32,33,Kenn Rein,Six,96,female
33,34,Gain Toe,Seven,69,male
34,35,Rows Noump,Six,88,female


In [ ]:
# head and tail methods will return 5 rows by default.

In [ ]:
# random sample
df.sample()

,id,name,class,mark,gender
5,6,Alex John,Four,55,male


In [ ]:
# 5 random samples
df.sample(5)

,id,name,class,mark,gender
20,21,Babby John,Four,69,female
1,2,Max Ruin,Three,85,male
27,28,Rojj Base,Seven,86,female
18,19,Tinny,Nine,18,male
4,5,John Mike,Four,60,female


In [ ]:
# number of rows and columns
df.shape
# It doesn't have (). Its and attribute (rows, columns)

(35, 5)

In [ ]:
# dataset information
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 35 entries, 0 to 34
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   id      35 non-null     int64 
 1   name    34 non-null     object
 2   class   34 non-null     object
 3   mark    35 non-null     int64 
 4   gender  33 non-null     object
dtypes: int64(2), object(3)
memory usage: 1.5+ KB




* look at the number of entries and compare it to Non-null count.
* if they look different, that column has a missing value.
* name column has 34. 1 name is missing.
* class column has a missing data.
* gender column has 33. 2 info missing.





* object data type is string.

In [ ]:
# name of the columns
df.columns

Index(['id', 'name', 'class', 'mark', 'gender'], dtype='object')

In [ ]:
# statistical summary of numeric columns only
df.describe()

,id,mark
count,35.000000,35.000000
mean,18.000000,74.657143
std,10.246951,16.401117
min,1.000000,18.000000
25%,9.500000,62.500000
50%,18.000000,79.000000
75%,26.500000,88.000000
max,35.000000,96.000000


* std = standard deviation
* 25%: 25% of the marks of the student are equal to or lesss than 62.5
* 50% median. 50% of the marks of the student are equal to or lesss than 79
* if std is close or greater than mean, there are outliers (possibly extreme data, remove or replace) in your data

In [ ]:
# number of rows
len(df)

35

In [ ]:
# rows * columns
df.size
# 35 * 5 = 175

175

In [ ]:
# datatypes of the columns
df.dtypes

,0
id,int64
name,object
class,object
mark,int64
gender,object


In [ ]:
# Counts missing values per column.
df.isnull().sum()

,0
id,0
name,1
class,1
mark,0
gender,2


In [ ]:
# Counts duplicate rows.
df.duplicated().sum()

np.int64(0)

# How to select rows or coloumns

## Subsetting DataFrame

In [ ]:
df["name"]
# slicing method

,name
0,John Deo
1,Max Ruin
2,Arnold
3,Krish Star
4,John Mike
5,Alex John
6,My John Rob
7,Asruid
8,Tes Qry
9,Big John


,name
0,John Deo
1,Max Ruin
2,Arnold
3,Krish Star
4,John Mike
5,Alex John
6,My John Rob
7,Asruid
8,Tes Qry
9,Big John


In [ ]:
df["name"].head()

,name
0,John Deo
1,Max Ruin
2,Arnold
3,Krish Star
4,John Mike


In [ ]:
df["class"].sample(3)

,class
1,Three
6,Fifth
8,Six


In [ ]:
df.name

,name
0,John Deo
1,Max Ruin
2,Arnold
3,Krish Star
4,John Mike
5,Alex John
6,My John Rob
7,Asruid
8,Tes Qry
9,Big John


In [ ]:
df.class
# this will return error because class is reserved key.

SyntaxError: invalid syntax (1101718610.py, line 1)

In [ ]:
df.gender.tail()

,gender
30,male
31,female
32,female
33,male
34,female


In [ ]:
# names and marks
df[["name", "mark"]]
# [[]]. The first [] is for slicking method. The second [] is list of columns.

,name,mark
0,John Deo,75
1,Max Ruin,85
2,Arnold,55
3,Krish Star,60
4,John Mike,60
5,Alex John,55
6,My John Rob,78
7,Asruid,85
8,Tes Qry,78
9,Big John,55


In [ ]:
df[["name", "mark"]].sample(3)

,name,mark
24,Giff Tow,88
4,John Mike,60
23,Tiddy Now,78


In [ ]:
df["mark"].min()

18

In [ ]:
df["mark"].median()

79.0

In [ ]:
df.mark.mean()

np.float64(74.65714285714286)

In [ ]:
df.mark.sum()

np.int64(2613)

## Exercise 1

In [ ]:
df[["name", "class", "gender"]]

,name,class,gender
0,John Deo,Four,female
1,Max Ruin,Three,male
2,Arnold,Three,male
3,Krish Star,Four,female
4,John Mike,Four,female
5,Alex John,Four,male
6,My John Rob,Fifth,male
7,Asruid,Five,male
8,Tes Qry,Six,NaN
9,Big John,Four,female


In [ ]:
df["mark"][0]

np.int64(75)

In [ ]:
df["mark"][3]

np.int64(60)

## loc and iloc

In [ ]:
# name of the student at first row
df.loc[0, "name"]
# 0 is the index number for the first row.

'John Deo'

In [ ]:
df.iloc[0, 1]
# 0 is the index number for the first row.
# 1 is the index number for "name" column.

'John Deo'

In [ ]:
# first 5 rows, id, name, gender
df.loc[0:4, ["id", "name", "gender"]]
# 0:4 is first 5 rows.
# these are list of the column names  ["id", "name", "gender"].
# loc operator includes stopping point of index.

,id,name,gender
0,1,John Deo,female
1,2,Max Ruin,male
2,3,Arnold,male
3,4,Krish Star,female
4,5,John Mike,female


In [ ]:
df.iloc[0:5, [0, 1, 4]]
# iloc operator does not include stopping point of index.

,id,name,gender
0,1,John Deo,female
1,2,Max Ruin,male
2,3,Arnold,male
3,4,Krish Star,female
4,5,John Mike,female


In [ ]:
df.loc[27, "class"]

'Seven'

In [ ]:
df.loc[:4, "name"]
# :4 starts from first index, print until index 4

,name
0,John Deo
1,Max Ruin
2,Arnold
3,Krish Star
4,John Mike


In [ ]:
df.loc[29: , "class"]
# 29: start from index 29 till the last

,class
29,Six
30,Four
31,Seven
32,Six
33,Seven
34,Six


In [ ]:
df.loc[11:16, ["name", "mark"]]

,name,mark
11,Recky,94
12,Kty,88
13,Bigy,88
14,Tade Row,88
15,Gimmy,88
16,Tumyu,54


In [ ]:
df.loc[[1, 7], "name"]

,name
1,Max Ruin
7,Asruid


In [ ]:
df.iloc[[1, 7], 1]

,name
1,Max Ruin
7,Asruid


In [ ]:
# id, class, name of first 3 students
df.loc[0:2, ["id", "class", "name"]]

,id,class,name
0,1,Four,John Deo
1,2,Three,Max Ruin
2,3,Three,Arnold


In [ ]:
df.iloc[0:3, [0, 2, 1]]

,id,class,name
0,1,Four,John Deo
1,2,Three,Max Ruin
2,3,Three,Arnold


In [ ]:
# average mark first 5 students
df.loc[0:4, "mark"].mean()

np.float64(67.0)

In [ ]:
df.iloc[0:5, 3].mean()

np.float64(67.0)

In [ ]:
df.loc[20:30, ["name", "gender"]].shape

(11, 2)

## How to filter (you cannot use iloc in filtering)

In [ ]:
df["mark"]>60
# boolen masking (true or false)

,mark
0,True
1,True
2,False
3,False
4,False
5,False
6,True
7,True
8,True
9,False


In [ ]:
df[df["mark"]>60]

,id,name,class,mark,gender
0,1,John Deo,Four,75,female
1,2,Max Ruin,Three,85,male
6,7,My John Rob,Fifth,78,male
7,8,Asruid,Five,85,male
8,9,Tes Qry,Six,78,NaN
10,11,Ronald,Six,89,female
11,12,Recky,Six,94,female
12,13,Kty,Seven,88,female
13,14,Bigy,Seven,88,female
14,15,Tade Row,NaN,88,male


In [ ]:
# df[condition]

In [ ]:
df[df["class"]=="Six"]

,id,name,class,mark,gender
8,9,Tes Qry,Six,78,NaN
10,11,Ronald,Six,89,female
11,12,Recky,Six,94,female
16,17,Tumyu,Six,54,male
29,30,Reppy Red,Six,79,female
32,33,Kenn Rein,Six,96,female
34,35,Rows Noump,Six,88,female


In [ ]:
# df[] slicing method
# df["class"]=="Six" filtering rows

In [ ]:
df[df["gender"]=="female"].head()

,id,name,class,mark,gender
0,1,John Deo,Four,75,female
3,4,Krish Star,Four,60,female
4,5,John Mike,Four,60,female
9,10,Big John,Four,55,female
10,11,Ronald,Six,89,female


In [ ]:
df[df["mark"]>90]

,id,name,class,mark,gender
11,12,Recky,Six,94,female
32,33,Kenn Rein,Six,96,female


In [ ]:
# female student at class four
# slicing method for multiple conditions >> df[(condition1) & (condition2)]
df[(df["gender"]=="female") & (df["class"]=="Four")]

,id,name,class,mark,gender
0,1,John Deo,Four,75,female
3,4,Krish Star,Four,60,female
4,5,John Mike,Four,60,female
9,10,Big John,Four,55,female
20,21,Babby John,Four,69,female


In [ ]:
# create report of student with their mark less than 70 at class seven
df[(df["mark"]<70) & (df["class"]=="Seven")]

,id,name,class,mark,gender
21,22,Reggid,Seven,55,female
28,29,Tess Played,Seven,55,male
33,34,Gain Toe,Seven,69,male


df.loc[condition, column name(s)]

In [ ]:
# df[df["gender"]=="female"].head()
# df.loc[rows, coloumn names]

df.loc[df["gender"]=="female", :].head()

# : return all columns

,id,name,class,mark,gender
0,1,John Deo,Four,75,female
3,4,Krish Star,Four,60,female
4,5,John Mike,Four,60,female
9,10,Big John,Four,55,female
10,11,Ronald,Six,89,female


In [ ]:
df.loc[df["gender"]=="female", ["name", "gender"]].head()

,name,gender
0,John Deo,female
3,Krish Star,female
4,John Mike,female
9,Big John,female
10,Ronald,female


In [ ]:
# df[df["mark"]>90]

df.loc[df["mark"]>90, :]

,id,name,class,mark,gender
11,12,Recky,Six,94,female
32,33,Kenn Rein,Six,96,female


- if you are using slicing method for a single column df[""]
- if you are using slicing method for a multiple columns df[["coloumn1", "coloumn2"]]
- ["coloumn1", "coloumn2"] is for list of columnn names
- filter data for single column df[df["mark"]>90]
- filter data for multipe column df[(condition1) & (condition2)]

- if you are using df.loc[] or df.iloc[]
- if you are using for filtering df.loc[df["mark">90], :]

In [ ]:
# df[(df["mark"]<70) & (df["class"]=="Seven")]
df.loc[(df["mark"]<70) & (df["class"]=="Seven")]

,id,name,class,mark,gender
21,22,Reggid,Seven,55,female
28,29,Tess Played,Seven,55,male
33,34,Gain Toe,Seven,69,male


### Filter method (very limited)

In [ ]:
df.filter(items=['name', 'mark']).head()

,name,mark
0,John Deo,75
1,Max Ruin,85
2,Arnold,55
3,Krish Star,60
4,John Mike,60


In [ ]:
df.filter(like='ma').head()

,mark
0,75
1,85
2,55
3,60
4,60


### Query method

In [ ]:
df.query("mark > 60").head()

,id,name,class,mark,gender
0,1,John Deo,Four,75,female
1,2,Max Ruin,Three,85,male
6,7,My John Rob,Fifth,78,male
7,8,Asruid,Five,85,male
8,9,Tes Qry,Six,78,NaN


In [ ]:
df.query('gender == "female"')

,id,name,class,mark,gender
0,1,John Deo,Four,75,female
3,4,Krish Star,Four,60,female
4,5,John Mike,Four,60,female
9,10,Big John,Four,55,female
10,11,Ronald,Six,89,female
11,12,Recky,Six,94,female
12,13,Kty,Seven,88,female
13,14,Bigy,Seven,88,female
19,20,Jackly,Nine,65,female
20,21,Babby John,Four,69,female


## Exercise 2

In [ ]:
# 1. Average mark across all students
df["mark"].mean()

np.float64(74.65714285714286)

In [ ]:
# 2. Average mark across male and average mark across female students
(df[df["gender"]=="male"]["mark"].mean(),
df[df["gender"]=="female"]["mark"].mean())

(np.float64(71.58823529411765), np.float64(77.3125))

In [ ]:
# 3. Max and min mark
df["mark"].max(),
df["mark"].min()

(96, 18)

In [ ]:
print(f"Max mark: {df['mark'].max()}")
print(f"Min mark: {df['mark'].min()}")

Max mark: 96
Min mark: 18


In [ ]:
# 4. Max mark for male and max mark for female students
(df[df["gender"]=="male"]["mark"].max(),
df[df["gender"]=="female"]["mark"].max())

(88, 96)

In [ ]:
# 5. Average mark for class 'Five'
df[df["class"]=="Five"]["mark"].mean()

np.float64(80.0)

In [ ]:
df.loc[df["class"]=="Five", "mark"].mean()

np.float64(80.0)

In [ ]:
# 6. Average mark for class 'Six' and male students only
df[(df["class"]=="Six") & (df["gender"]=="male")]["mark"].mean()

np.float64(54.0)

## Sort values

In [ ]:
df.sort_values(by="mark").head()

,id,name,class,mark,gender
18,19,Tinny,Nine,18,male
16,17,Tumyu,Six,54,male
2,3,Arnold,Three,55,male
5,6,Alex John,Four,55,male
28,29,Tess Played,Seven,55,male


In [ ]:
df.sort_values(by="name").head()

,id,name,class,mark,gender
5,6,Alex John,Four,55,male
2,3,Arnold,Three,55,male
7,8,Asruid,Five,85,male
20,21,Babby John,Four,69,female
9,10,Big John,Four,55,female


In [ ]:
df["mark"].sort_values().head()

,mark
18,18
16,54
2,55
5,55
28,55


In [ ]:
df["name"].sort_values().head(3)

,name
5,Alex John
2,Arnold
7,Asruid


In [ ]:
df.loc[0:4, ["name", "mark"]].sort_values(by="mark", ascending=False)
# ascending=False will sot in descending order

,name,mark
1,Max Ruin,85
0,John Deo,75
3,Krish Star,60
4,John Mike,60
2,Arnold,55


In [ ]:
# find male student with their mark higher than 70 and sort by their class
df.loc[(df["gender"]=="male") & (df["mark"]>70)].sort_values(by="class")

,id,name,class,mark,gender
22,23,Herod,Eight,79,male
6,7,My John Rob,Fifth,78,male
7,8,Asruid,Five,85,male
17,18,Honny,Five,75,male
15,16,Gimmy,Four,88,male
30,31,Marry Toeey,Four,88,male
23,24,Tiddy Now,Seven,78,male
24,25,Giff Tow,Seven,88,male
25,26,Crelea,Seven,79,male
1,2,Max Ruin,Three,85,male


In [ ]:
# Write code to find information of female students who are above the classroom average. Sort the results in descening order.
df[(df["gender"]=="female") & (df["mark"]>df["mark"].mean())].sort_values(by="mark", ascending=False)

,id,name,class,mark,gender
32,33,Kenn Rein,Six,96,female
11,12,Recky,Six,94,female
31,32,Binn Rott,Seven,90,female
10,11,Ronald,Six,89,female
34,35,Rows Noump,Six,88,female
12,13,Kty,Seven,88,female
13,14,Bigy,Seven,88,female
27,28,Rojj Base,Seven,86,female
29,30,Reppy Red,Six,79,female
0,1,John Deo,Four,75,female


In [ ]:
# How many male and female students are there?
df["gender"].value_counts()

,count
gender,
male,17
female,16


In [ ]:
# Class and number of student
df["class"].value_counts()

,count
class,
Seven,10
Four,8
Six,7
Three,3
Nine,2
Five,2
Fifth,1
Eight,1
